domain/secretbackend/state/state.go

```
@@ -0,0 +1,562 @@
+		if err != nil {
+			return fmt.Errorf("querying secret backends: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot list secret backends: %w", err)
+	}
+	if len(rows) == 0 {
+		return nil, errors.NotFoundf("secret backend %q", v)
+	}
+	return rows.ToSecretBackendInfo()[0], errors.Trace(err)
+}
+
+// GetSecretBackendByName returns the secret backend for the given backend name.
+func (s *State) GetSecretBackendByName(ctx context.Context, backendName string) (*coresecrets.SecretBackend, error) {
+	return s.getSecretBackend(ctx, "name", backendName)
+}
+
+// GetSecretBackend returns the secret backend for the given backend ID.
+func (s *State) GetSecretBackend(ctx context.Context, backendID string) (*coresecrets.SecretBackend, error) {
+	return s.getSecretBackend(ctx, "uuid", backendID)
+}
+
+// SecretBackendRotated updates the next rotation time for the secret backend.
+func (s *State) SecretBackendRotated(ctx context.Context, backendID string, next time.Time) error {
+	if _, err := s.GetSecretBackend(ctx, backendID); err != nil {
+		// Check if the backend exists or not here because the
+		// below UPDATE operation won't tell us.
+		return errors.Trace(err)
+	}
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	stmt, err := sqlair.Prepare(`
+UPDATE secret_backend_rotation
+SET next_rotation_time = $SecretBackendRotation.next_rotation_time
+WHERE backend_uuid = $SecretBackendRotation.backend_uuid
+    AND next_rotation_time > $SecretBackendRotation.next_rotation_time`, SecretBackendRotation{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt,
+			SecretBackendRotation{
+				BackendUUID:      backendID,
+				NextRotationTime: sql.NullTime{Time: next, Valid: true},
+			},
+		).Run()
+		if err != nil {
+			return fmt.Errorf("updating secret backend rotation: %w", err)
```

We should return a SecretBackendNotFound error if necessary here, same in other places

---

It's done in the 1st line of this method.

---

But that's outside the txn. So it can lead to errors.
We need to do the check inside the txn